Improved inactivity timeout workflow

orangejenny opened the pull request:
##### SUMMARY

https://trello.com/c/yNP02MmM/39-inactivity-timer-improvements

Followup for https://github.com/dimagi/commcare-hq/pull/27572

This adds two modals dealing with inactivity timeouts. The first pops up 2 minutes before a session will expire, with a warning and an option to extend the session another 30 min (or however long the timeout is). The second pops up when the user is actually logged out and lets them re-log in. This second modal cannot be closed until the user is genuinely logged in.

This works by adding the timestamp of the last request to the user's session and then polling the server to see if the user is still logged in around the time the session should expire. This does not affect the actual expiration of sessions. If the modal doesn't appear for whatever reason, the user still gets logged out. Also, pages that aren't domain-specific (account settings, admin pages, etc.) don't use this workflow.

##### FEATURE FLAG
This only affects domains that use the "Shorten Inactivity Timeout" setting.

##### RISK ASSESSMENT / QA PLAN
[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-1331)

##### PRODUCT DESCRIPTION
Warning modal:
![Screen Shot 2020-05-27 at 1 15 28 PM](https://user-images.githubusercontent.com/1486591/83059409-da9b4900-a027-11ea-964c-d3e37340d0a2.png)

Modal on logout:
![Screen Shot 2020-05-27 at 1 50 06 PM](https://user-images.githubusercontent.com/1486591/83059448-e850ce80-a027-11ea-8eea-468087364ea6.png)

Then 2fa screen, if applicable:
![Screen Shot 2020-05-27 at 2 07 27 PM](https://user-images.githubusercontent.com/1486591/83059521-074f6080-a028-11ea-8bba-40699e994f68.png)

Then success screen directing user to close the modal:
![Screen Shot 2020-05-27 at 1 50 46 PM](https://user-images.githubusercontent.com/1486591/83059552-12a28c00-a028-11ea-8bf9-42cf71b8a920.png)

orangejenny committed:
Added modal

orangejenny committed:
Added iframe with login

orangejenny committed:
Added ping_login view

orangejenny committed:
Added polling

orangejenny committed:
Updated delay to match domain's secure_timeout

orangejenny committed:
Moved login_new_window to hqwebapp and removed domain parameter

orangejenny committed:
Updated iframe size to take up entire modal

orangejenny committed:
Updated to not display modal on certain pages

Don't display on app preview or on login flow.

orangejenny committed:
Added logic to ensure authentication before dismissing modal

orangejenny committed:
Restored last_request value in session

This was removed in https://github.com/dimagi/commcare-hq/pull/24140/ with the switch to Django's set_expiry.
Using set_expiry and django.contrib.sessions.backends.cache means we can't determine when a session is set
to expire, so add that value back here.

orangejenny committed:
Added a few more URLs that don't count as activity

orangejenny committed:
Extracted js functions

orangejenny committed:
Updated poll frequency

orangejenny committed:
Added modal warning when session is close to ending

orangejenny committed:
Added iframe-friendly login view

orangejenny committed:
Added verification that new user is same as old user

orangejenny commented:
fyi @dimagi/product

millerdev commented on corehq/apps/hqwebapp/static/hqwebapp/js/inactivity.js:
```diff
@@ -0,0 +1,115 @@
+) {
+    $(function () {
+        var timeout = initialPageData.get('secure_timeout') * 60 * 1000,    // convert from minutes to milliseconds
+            $modal = $("#inactivityModal"),     // won't be present on app preview or pages without a domain
```
Can these comments be verified with tests?

orangejenny replied:
The way I'd test it would be to call the view and then check the generated HTML with a regex for this id, which seems like a slow and fragile for a fairly minor test case. Any thoughts on a better way to test this?

millerdev reviewed (approved):
> Then success screen directing user to close the modal:

Could it close automatically, allowing them to get back to work quicker?

Is there any chance that this could cause a bad UX if the person is typing when the warning modal pops up? Maybe wait a few seconds after the last keypress before showing the modal if the person is typing and there is ample time left before expiry?

Tests would be nice. Ideally Python tests would verify the workflow on the server-side (check things like `ping_login` does not extend session, `bsd_license` (`extendSession`) does extend session, etc.) and JS tests would verify client experience (modal appears as expected, modal automatically transforms from "extend session" to "please re-login", etc).

orangejenny committed:
Updated templatetags test

orangejenny committed:
Fixed lint errors

orangejenny committed:
Renamed initial page data key

Web Apps already has a 'username' key, so they conflict.

orangejenny committed:
Extracted warning modal functions and added focus on modal open

orangejenny committed:
Updated warning modal to not show while user is actively typing

orangejenny committed:
Added logic to automatically hide modal

orangejenny commented:
Hey, thanks for the thoughtful feedback.

> Could it close automatically, allowing them to get back to work quicker?

I wrote out a long explanation of why this would be nice but too hard, and then I gave it a shot and found it's easy to just check authorization each time the iframe loads a new url. e03d23b

> Is there any chance that this could cause a bad UX if the person is typing when the warning modal pops up? Maybe wait a few seconds after the last keypress before showing the modal if the person is typing and there is ample time left before expiry?

This is a great idea, and Sarah also brought it up. Added in 9ccdc43

> Tests would be nice. Ideally Python tests would verify the workflow on the server-side (check things like ping_login does not extend session, bsd_license (extendSession) does extend session, etc.) and JS tests would verify client experience (modal appears as expected, modal automatically transforms from "extend session" to "please re-login", etc).

I'll take a look at adding python tests, possibly in a followup PR. I'm not likely to add js tests, I think they'll be a lot of effort and likely fragile. We do not have a good test setup for js that interacts with the DOM.

orangejenny committed:
Added specific message when user logs in as different user

orangejenny committed:
Fixed error handling

millerdev reviewed (approved):
I think the answer is "no", but any chance the polling will cause a thundering herd of requests that negatively impacts our server performance?

orangejenny commented:
> a thundering herd of requests

I've been thinking about this and am currently not too worried. The frequent polling (every 10 sec or less) is only when a user is within 2 minutes of their session expiring and stops once the session does expire.

orangejenny commented on corehq/apps/hqwebapp/static/hqwebapp/js/inactivity.js:
```diff
@@ -118,6 +124,17 @@ hqDefine('hqwebapp/js/inactivity', [
+            if (warningActive) {
+                showWarningModal();
+            }
+        }, 500));
```
I had not seen that SO answer. 1 sec is about how long it takes to move your hand from mouse to keyboard or vice versa, which is a much bigger gesture than moving between keyboard keys. So I'm estimating that 500ms is enough of a delay that you're probably not steadily typing. I did some testing (console logging when HQ thought I was or wasn't typing) and it felt natural enough, though of course typing speed varies. I don't want the delay to be so long that the user possibly doesn't see the warning at all before they get logged out, since that's much more disruptive than the warning. I'd be up for increasing it to 1 sec or maybe 2, but probably not more than that.

orangejenny committed:
Added prefillled email address to logiin popup

orangejenny committed:
Merge pull request #27659 from dimagi/jls/relogin-prefill

Prefill username on login popup

orangejenny committed:
Added closing of select2s

orangejenny committed:
Stopped warning from showing if user is moving mouse

orangejenny committed:
Updated domain reference

To my surprise, it's possible for request.project to be present but domain
to be blank. This can happen while logging in as a mobile user.

orangejenny committed:
Merge pull request #27660 from dimagi/jls/relogin-fixes

Minor additions to relogin workflow

esoergel commented on corehq/apps/hqwebapp/views.py:
```diff
@@ -447,6 +448,12 @@ def domain_login(req, domain, custom_template_name=None, extra_context=None):
     return _login(req, domain, custom_template_name, extra_context)
 
 
+def iframe_login(req, domain):
```
Do these views need to be marked location-safe?

esoergel replied:
nit: I think this should be `iframe_domain_login`, as it's a variant of `domain_login`, not `login`

orangejenny replied:
4f884724a5bee18b0dd64af9a3c03211721ca6c3

orangejenny replied:
and a81f6e211b81bcdab250b8b5303e9e948ae24972

esoergel commented on corehq/apps/hqwebapp/views.py:
```diff
@@ -447,6 +448,12 @@ def domain_login(req, domain, custom_template_name=None, extra_context=None):
+def iframe_login(req, domain):
+    return domain_login(req, domain, custom_template_name="hqwebapp/iframe_login.html", extra_context={
+        'current_page': {'page_name': _('Your session has expired')},
+    })
```
I see this was already like this, but FYI it looks like `extra_context` is implemented in such a way that doing this would override any other variables in the `'current_page'` dict if one already exists:
```python
In [1]: ctxt = {'foo': 'bar',
   ...:         'current_page': {'page_name': 'name1',
   ...:                          'url': 'example.com'}}

In [2]: ctxt
Out[2]: {'current_page': {'page_name': 'name1', 'url': 'example.com'}, 'foo': 'bar'}

In [3]: ctxt.update({'current_page': {'page_name': 'name2'}})

In [4]: ctxt
Out[4]: {'current_page': {'page_name': 'name2'}, 'foo': 'bar'}  # 'url' is missing
```

orangejenny replied:
I'm going to leave it as-is, as the current usages of `extra_context` aren't affected by this, but that is a good point.
